domain/externalcontroller/state/state.go

```diff
@@ -59,125 +65,235 @@ WHERE  	ctrl.uuid = $M.id`
 	db, err := st.DB()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
+	// TODO(nvinuesa): We should use an `IN` statement query here, instead
+	// of looping over the list of models and performing N queries, but
+	// they are not yet supported on sqlair:
+	// https://github.com/canonical/sqlair/pull/76
 	q := `
-SELECT 	(ctrl.uuid, alias, ca_cert, address) as &ExternalController.* 
-FROM   	external_controller AS ctrl	
-	JOIN external_model AS model
-	ON ctrl.uuid = model.controller_uuid
-       	LEFT JOIN external_controller_address AS addrs
-       	ON ctrl.uuid = addrs.controller_uuid
-WHERE  	model.uuid = $M.id`
+SELECT (ctrl.uuid,  
+       ctrl.alias,
+       ctrl.ca_cert,
+       addrs.address) as &ExternalController.*,
+       model.uuid as &ExternalController.model
+FROM external_controller AS ctrl	
+JOIN external_model AS model
+ON ctrl.uuid = model.controller_uuid
+LEFT JOIN external_controller_address AS addrs
+ON ctrl.uuid = addrs.controller_uuid
+WHERE model.uuid = $M.model`
+
+	var resultControllers []crossmodel.ControllerInfo
+
+	uniqueControllers := make(map[string]crossmodel.ControllerInfo)
 	s, err := sqlair.Prepare(q, ExternalController{}, sqlair.M{})
 	if err != nil {
 		return nil, errors.Annotatef(err, "preparing %q", q)
 	}
 
 	var rows []ExternalController
 	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		return errors.Trace(tx.Query(ctx, s, sqlair.M{"id": modelUUID}).GetAll(&rows))
-	}); err != nil {
-		return nil, errors.Annotate(err, "querying external controller for model")
-	}
+		for _, modelUUID := range modelUUIDs {
+			err := tx.Query(ctx, s, sqlair.M{"model": modelUUID}).GetAll(&rows)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if len(rows) > 0 {
+				uniqueControllers[rows[0].ID] = *controllerInfoFromRows(rows)
+			}
 
-	if len(rows) == 0 {
-		return nil, errors.NotFoundf("external controller for model %q", modelUUID)
+			for _, controller := range uniqueControllers {
+				resultControllers = append(resultControllers, controller)
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying external controller")
 	}
 
-	return controllerInfoFromRows(rows[0].ID, rows), nil
+	return resultControllers, nil
 }
 
-func controllerInfoFromRows(uuid string, rows []ExternalController) *crossmodel.ControllerInfo {
-	// We know that we queried for a single ID, so the first instance
-	// of the controller info fields will be repeated.
-	ci := crossmodel.ControllerInfo{
-		ControllerTag: names.NewControllerTag(uuid),
-		Alias:         rows[0].Alias.String,
+// controllerInfoFromRows is a convenient method for de-normalizing multiple
+// rows containing the (necessary) columns from the three tables that
+// represent the ControllerInfo struct (i.e. including addresses and models).
+func controllerInfoFromRows(rows []ExternalController) *crossmodel.ControllerInfo {
```

Comment:
This gets deleted.

Reply:
done, see 95564b14e4c4efb7653904e05c7eeab5ab1bbe9b